08-IO/Party.hs

```diff
@@ -0,0 +1,181 @@
+module Party where
+
+import Data.Monoid
+import Data.Semigroup
+import Data.Tree
+import Employee
+
+--Exercise 1
+{-
+A function
+glCons :: Employee -> GuestList -> GuestList
+which adds an Employee to the GuestList (updating the cached
+Fun score appropriately). Of course, in general this is impossible:
+the updated fun score should depend on whether the Employee
+being added is already in the list, or if any of their direct subordinates are in the list, and so on. For our purposes, though, you
+may assume that none of these special cases will hold: that is,
+glCons should simply add the new Employee and add their fun
+score without doing any kind of checks.
+-}
+glCons :: Employee -> GuestList -> GuestList
+glCons emp (GL emps funScore) = GL (emp : emps) (empFun emp + funScore)
+
+
+{-
+A Monoid instance for GuestList.
+(How is the Monoid instance 2
+supposed to work, you ask? You figure it out!)
+-}
+instance Monoid GuestList where
+    mempty = GL [] 0
+
+instance Semigroup GuestList where
+    (GL emps1 fun1) <> (GL emps2 fun2) = GL (emps1 ++ emps2) (fun1 + fun2)
+
+{-
+A function moreFun :: GuestList -> GuestList -> GuestList
+which takes two GuestLists and returns whichever one of them
+is more fun, i.e. has the higher fun score. (If the scores are equal it
+does not matter which is returned.)
+-}
+moreFun :: GuestList -> GuestList -> GuestList
+moreFun gl1@(GL _ fun1) gl2@(GL _ fun2)
+    | fun2 > fun1 = gl2
+    | otherwise   = gl1
+
+
+{-
+Exercise 2
+The Data.Tree module from the standard Haskell libraries defines
+the type of “rose trees”, where each node stores a data element and
+has any number of children (i.e. a list of subtrees):
+data Tree a = Node {
+rootLabel :: a, -- label value
+subForest :: [Tree a] -- zero or more child trees
+}
+Strangely, Data.Tree does not define a fold for this type! Rectify the
+situation by implementing
+treeFold :: ... -> Tree a -> b
+(See if you can figure out what type(s) should replace the dots in
+the type of treeFold. If you are stuck, look back at the lecture notes
+from Week 7, or infer the proper type(s) from the remainder of this
+assignment.)
+-}
+--Exercise 2
+treeFold :: (a -> [b] -> b) -> b -> Tree a -> b
+treeFold f init (Node val children) = f val (map (treeFold f init) children)
```

Comment:
As an exercise, you can try implementing a strict fold, analagous to `foldl'`.

Reply:
@kwannoel
I'm having some trouble coming up with an implementation with the same type signature as `treeFold`. I have the below implementation which is slightly tweaked in the signature of `f`. Maybe you can help me out here?
```haskell
treeFold' :: (b -> a -> b) -> b -> Tree a -> b
treeFold' f init (Node val [])       = f init val
treeFold' f init (Node val children) = 
    let strictAcc = f init val
    in seq strictAcc $ foldl' (treeFold' f) (strictAcc) children
```

Reply:
Your implementation is correct. It is complement to `foldr`.

See the type signature of `foldr` here: https://hackage.haskell.org/package/base-4.12.0.0/docs/Data-Foldable.html#t:Foldable

The differing implementation is your `treeFold`. Most people define as `treeFold :: (a -> b -> b) -> b -> Tree a -> b`, as shown in `Data.Foldable`.

Your `treeFold` is a *Breadth-first fold* rather than accumulating, node-wise. As such there's a difference of behaviour.

In most cases, we want to describe node-wise folding, this means we can just deal with 2 elements.
With bf folding, as you can see from your type signature: `f :: a -> [b] -> b`, we need to deal `a` and a collection, `[b]`, which usually means using more brainpower :(

Of course there are usecase for bf folding, but im not sure if that's what you need in this particular case.

Now in the event you want to perform strict bf fold:

```haskell
treeFoldStrict :: (b -> [a] -> b) -> b -> Tree a -> b
treeFoldStrict f init (Node val children) =
    seq acc $ treeFoldChildren f acc children
    where
        acc = f init [val]

treeFoldChildren :: (b -> [a] -> b) -> b -> [Tree a] -> b
treeFoldChildren f b children =
    seq acc $ treeFoldChildren f acc (concat branches)
    where
        acc = f b childValues
        childValues = fmap getNodeValue children
        branches = fmap getBranches children
        getNodeValue (Node v _) = v
        getBranches (Node _ b) = b
```

Reply:
https://gist.github.com/kwannoel/4a73a4bb0455a3a460a852b9bd21f312

Reply:
@kwannoel awesome thanks 👍

Reply:
I think you don't need function AND a value to implement proper fold for the tree.
For example the canonical foldr function for list takes `(a -> b -> b)` which tells you what to replace `(:)` constructor with and a `b` which tells what to replace `[]` constructor with.

Since Tree only has one constructor, you should only need `(a -> [b] -> b)` to know how to fold the `Node` constructor.
Not sure if that's in alignment with these exercises, but I'd implement it as `treeFold :: (a -> [b] -> b) -> Tree a -> b`

Reply:
@jhrcek Thanks, yes I think you are right. I have never thought about the type signature of `fold` in terms of constructors before and didn't know I could as well. Thanks for bringing this up to me.

Reply:
This is the "canonical way" to do it, there's a good explanation of this idea e.g. here: https://stackoverflow.com/questions/16426463/what-constitutes-a-fold-for-types-other-than-list